### ISSUE TITLE
twisted.mail.test.test_smtp.TimeoutTestCase.testSMTPClient fails intermittently due to timing issues

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2832]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#2832|
|Type|defect|
|Created|2007-09-12 14:55:56Z|

The tests in TimeoutTestCase in this module use the system clock, which means they cannot be deterministic.  They should use a deterministic clock instead.

Attachments:

* [2832.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/91a/91a7bc9deaeeb18d77904dce1ed85ab89013e06f/b33bb2602935f2d21f8049957ea2b32f3d935a63.diff) (2930 bytes) - added by therve on 2007-10-30 17:02:25Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__2832 2832
type__defect defect
reporter__exarkun exarkun
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__mail mail
keywords__ 
time__1189608956000000 1189608956000000
changetime__1194557599000000 1194557599000000
version__None None
owner__ 
cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve commented|
|-|-|

(In [21533]) Remove time related calls.

Refs [#2832](https://github.com/twisted/twisted/issues/2832)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve removed owner|
|-|-|

Ready to review in test-smtp-timeout-2832.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to @therve|
|-|-|

That twisted.internet import line in test_smtp is getting kinda long. :)

Maybe the assertion about length of calls being 1 would be better in a callback on d, instead of having list.append as the callback.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve commented|
|-|-|

I don't understand. Do you mean removing the assertion? Because I don't see how it can be done in the callback, since it's asserting that the callback has been called? Maybe I can check the clock time in the callback?

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to @therve|
|-|-|

Sorry, yea, I phrased that poorly.  Instead of using ```calls.append``` as a callback, define a function which makes an assertion about ```clock.seconds()``` and use that as the callback (as you said).

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve commented|
|-|-|

(In [21627]) Process review comments.

Refs [#2832](https://github.com/twisted/twisted/issues/2832)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve commented|
|-|-|

(In [21631]) Return deferreds from assertFailure

Refs [#2832](https://github.com/twisted/twisted/issues/2832)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_10"></a>|@therve commented|
|-|-|

Sorry the last commit is irrelevant here.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_11"></a>|@therve commented|
|-|-|

(In [21640]) Some docs and comments.

Refs [#2832](https://github.com/twisted/twisted/issues/2832)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun set owner to @therve|
|-|-|

Looks good

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_14"></a>|@therve set status to `closed`|
|-|-|

(In [21666]) Merge test-smtp-timeout-2832

Author: therve
Reviewer: exarkun
Fixes [#2832](https://github.com/twisted/twisted/issues/2832)

Use a deterministic clock in twisted.mail.test.test_smtp.TimeoutTestCase to
avoid intermittent failures.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
